Quadriga/src/main/java/edu/asu/spring/quadriga/web/config/SimpleSignInAdapter.java

```diff
@@ -0,0 +1,110 @@
+import edu.asu.spring.quadriga.exceptions.QuadrigaStorageException;
+import edu.asu.spring.quadriga.exceptions.UsernameExistsException;
+import edu.asu.spring.quadriga.service.IUserManager;
+import edu.asu.spring.quadriga.service.QuadrigaUserDetails;
+import edu.asu.spring.quadriga.web.login.QuadrigaGrantedAuthority;
+import edu.asu.spring.quadriga.web.login.RoleNames;
+import edu.asu.spring.quadriga.web.manageusers.beans.AccountRequest;
+
+/**
+ * 
+ * This class authorizes (assigns appropriate roles) users authenticated by Github.
+ * @author Chiraag Subramanian 
+ *
+ */
+
+public final class SimpleSignInAdapter implements SignInAdapter {
+    
+    private final Logger logger = LoggerFactory.getLogger(SignInAdapter.class);
+    
+    private IUserManager userManager;
+    private IUserHelper userHelper;
+
+    public SimpleSignInAdapter(IUserManager userManager, IUserHelper userHelper) {
+        this.userManager = userManager;
+        this.userHelper = userHelper;
+    }
+
+    /**
+     * This method checks if Quadriga contains details of the user authenticated by Github.
+     * If the user details are present (which indicates the user is an authorized user), appropriate roles are assigned,
+     * Else if the user details are not present (which indicates the user is an unregistered / not approved user, restricted access is assigned)
+     */
+    
+    @Override
+    public String signIn(String userId, Connection<?> connection,
+            NativeWebRequest request) {
+        List<GrantedAuthority> authorities = new ArrayList<>();
+
+        IUser user = null;
+            user = (IUser)userManager.findUserByProviderUserId(connection.getKey().getProviderUserId(), connection.getKey().getProviderId());
+      
+        // if user details is not present in the database, create a database record for the user.
+        if (user == null) {
+            user = userHelper.createUser(connection);
+            AccountRequest accountRequest = new AccountRequest();
+            accountRequest.setUsername(user.getUserName());
+            accountRequest.setName( user.getName());
+            accountRequest.setEmail(user.getEmail());
+            accountRequest.setSocialSignIn(true);
+            accountRequest.setProvider(user.getProvider());
+            accountRequest.setUserIdOfProvider(user.getUserIdOfProvider());
+            try {
+                //userManager.addSocialUser(user.getUserName(), user.getName(), user.getEmail(), user.getProvider(), user.getUserIdOfProvider());
+                userManager.addNewUser(accountRequest);
+            } catch (QuadrigaStorageException e) {
+                logger.error("Could not add user.", e);
+            } catch (UsernameExistsException e) {
+                logger.error("Could not add user.", e);
+            } catch (QuadrigaNotificationException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+            }
```

Comment:
The code tries to save the user's account request in the database. The signIn method signature in org.springframework.social.connect.web.SignInAdapter does not allow us to throw exceptions. We cannot use ModelAndView to return an error message to the user, because the signIn method signature has the return type "String".

The method invocation "userManager.addNewUser(accountRequest)" may lead to exceptions in the following scenarios:
i) There exists a user in 'tbl_quadriga_user' with the username that matches the username of the new user (throws UsernameExistsException)
ii) There already exists a request in 'tbl_quadriga_user_requests' for the new user (throws UsernameExistsException)
iii) Storage Failure 

The scenario (i) will not occur wrt SocialSignIn, because the method call userHelper.createUser(connection); will user that the user will have a unique username.

The scenario (ii) will occur wrt SocialSignIn, when the system tries to place a new account request for the user even though an account requests already exists for the user.  This can be avoided by placing additional checks in the code.

The scenario (iii) will occur wrt SocialSignIn due to storage exception. In this case, the user's account request is not placed in the system. However, when the user tries to login after a while (and the database is working fine), the user's account request will be eventually placed in the system.

Is there a way we can handle the error scenario when the user cannot be stored in the database?

Reply:
You make some wrong assumptions here and your conclusion is wrong as well:

- since you're in a multithreaded environment, scenario (i) can happen (even though very unlikely); just think it through with two new users trying to make an account request at the same time.
- scenario (iii) the way you describe it, is not a very user-friendly way of dealing with the issue. If the user thinks they placed an account request, they might just wait for their account to be activated, which would never happen, since the account request was never created. Without a error notification, this is just a bad user experience.

- yes, the method is returning a string. This doesn't mean you can't give feedback to the user. Do you understand what the method is returning and what it does before returning?